### DISCUSSION TITLE
RPC failed; curl 56 GnuTLS recv error (-9) when running git clone

### DISCUSSION THREAD

**Discussion by ja-Bertolin** [community/community#134430] [New to GitHub] [answered]
I use to use git pull just for updating a repository (https://github.com/jekhokie/raspberry-noaa-v2.git) but recently I started to have problems withe the process (fatal: refusing to merge unrelated histories) so I decided to ejecute a git clone for starting from scratch

I am using Raspberri Pi 3B+ with bullseye

Now when I run git clone it complains with the messages:

Cloning into 'raspberry-noaa-v2'...
remote: Enumerating objects: 240, done.
remote: Counting objects: 100% (240/240), done.
remote: Compressing objects: 100% (216/216), done.
error: RPC failed; curl 56 GnuTLS recv error (-9): Error decoding the received TLS packet.
error: 6269 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: index-pack failed

I have been searching in internet but not clear the solution as it is quite generic. I spoke with the developer of the repository and he told me that git fails from time to time, however, I ran the same git clone in my Macbook: BigSur 11.7.10 and it works fine all the time so I suspect that RPi have something that does not work fine now (some months ago it did it ok )

Any tip or recommendation? Thanks

**Comment by jvcByte** [comment]
This issue can be caused by several factors, including network instability, issues with the Git version on your Raspberry Pi, or even the configuration of the SSL/TLS library. Here are some steps you can take to troubleshoot and potentially resolve the issue:

1. **Check Network Connection**:
   - Ensure your Raspberry Pi has a stable internet connection. You can test the connection with commands like `ping google.com` to see if there are any drops or delays.

2. **Increase Buffer Size**:
   - Sometimes increasing the buffer size can help with large repositories. Try running:
     ```bash
     git config --global http.postBuffer 524288000
     ```

3. **Update Git**:
   - Ensure you are using the latest version of Git. You can update Git on your Raspberry Pi with the following commands:
     ```bash
     sudo apt update
     sudo apt install git
     ```

4. **Use SSH Instead of HTTPS**:
   - Switching to SSH might bypass some of the issues related to HTTPS. You can clone the repository using SSH:
     ```bash
     git clone git@github.com:jekhokie/raspberry-noaa-v2.git
     ```

5. **Check TLS/SSL Libraries**:
   - Ensure that your SSL/TLS libraries are up to date. You can update them with the following commands:
     ```bash
     sudo apt update
     sudo apt upgrade
     sudo apt install --reinstall gnutls-bin
     ```

6. **Temporary Directory**:
   - Sometimes issues with the temporary directory can cause problems. You can change the temporary directory for Git with the following commands:
     ```bash
     export TMPDIR=/path/to/large/temp/dir
     git clone https://github.com/jekhokie/raspberry-noaa-v2.git
     ```

7. **Disable GnuTLS**:
   - As a last resort, you can try disabling GnuTLS and using OpenSSL instead by recompiling Git with OpenSSL support.

8. **Use a Shallow Clone**:
   - If the repository is large, a shallow clone might help:
     ```bash
     git clone --depth 1 https://github.com/jekhokie/raspberry-noaa-v2.git
     ```

Try these steps one by one and see if any of them resolve the issue. If the problem persists, it might be worth looking into more detailed network diagnostics or even considering a fresh installation of the operating system on your Raspberry Pi.

**Reply by ja-Bertolin** [reply]
Thanks John

I tried all your suggestions and it is still failing.

I have done the same git clone with another RPi (same OS) and it works fine (same network)

I have uninstalled git and re-install it but not success. I would like to avoid re-burn the SD with a fresh OS because I have 27Gb of images stored in the card so I have to find a way to make a copy of all of that.

Still working on that

**Reply by jvcByte** [reply]
I'm quite surprised non of the above worked.
on most occasion, it's either proxy or HTTP request but switching to SSH should solve the issue.

**Reply by ja-Bertolin** [reply]
When I move to SSH I get the following

git clone git@github.com:jekhokie/raspberry-noaa-v2.git
Cloning into 'raspberry-noaa-v2'...
Warning: Permanently added the ECDSA host key for IP address '[140.82.121.35]:443' to the list of known hosts.
git@ssh.github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

I have been reading information about "git@ssh.github.com: Permission denied (publickey).” message but not able to find a way to fix it

**Reply by jvcByte** [reply, marked as answer]
## This means the SSH key isn't being accepted by GitHub. The below steps should definitely solve your problem:

### 1. Check for Existing SSH Keys:
```bash
ls -al ~/.ssh
```
There should be files named `id_rsa` (private key) and `id_rsa.pub` (public key) or similarly named key pairs.
If there is no SSH key available you will receive a message like: `ls: cannot access '/c/Users/dell/.ssh': No such file or directory`
### 2. Generate a New SSH Key (if there is no SSH available else skip No. 2)
```bash
ssh-keygen -t rsa -b 4096 -C "your_email@example.com"
```
Follow the prompts properly to save the key, and you can press Enter to accept the default file location. When prompted for a passphrase, you can choose to set one or leave it empty by pressing Enter.
### 3. Type the below command to add the SSH Key to the SSH Agent
```bash
eval "$(ssh-agent -s)"
```
```bash
ssh-add ~/.ssh/id_rsa
```
### 4. Add Your SSH Key to Your GitHub Account
Copy your SSH key to your clipboard:
```bash
cat ~/.ssh/id_rsa.pub
```
Go to [GitHub SSH settings](https://github.com/settings/keys), click "New SSH key," paste your key, and save it.
### 5. Test your SSH connection to GitHub:
```bash
ssh -T git@github.com
```
You should see a success message like:
```
Hi username! You've successfully authenticated, but GitHub does not provide shell access.
```
### 6. Update the Remote URL.
Ensure your repository's remote URL uses the correct SSH format. Run:
```bash
git remote -v
```
If it shows a URL starting with `https://`, update it to the SSH format:
```bash
git remote set-url origin git@github.com:jekhokie/raspberry-noaa-v2.git
```
### 7. Try cloning the repository again:
```bash
git clone git@github.com:jekhokie/raspberry-noaa-v2.git
```
This will resolve the `Permission denied (publickey)` issue and successfully clone your repository via SSH.

**Reply by ja-Bertolin** [reply]
Hi John, thanks a lot for you superb support, it worked like a charm

I perform all your steps and I was able to clone the repository and git pull (after installation correctly) works too.

I understand that once installed via SSH , there is not problem to run git pull without any other consideration.

So, I marked as Answer

Much appreciated.

**Reply by jvcByte** [reply]
It's a pleasure.

**Comment by WaqarAhmadWD** [comment]
if you clone a repo using HTTP then is error may be shown. i was also facing but i found the solution. switch to SSH it will work fine with large projects and seems very relaible.

**Reply by HarimbolaSantatra** [reply]
So switching to SSH solved the issue for me. It was a TLS related problem but why ?

**Comment by TarikIbrahim** [comment]
`sudo apt install --reinstall gnutls-bin`
installing the package solved the problem
